Make full use of UseCase class

Bachibouzouk opened the pull request:
Fix #83 
Fix #81
Closes #35 
Closes #58  

Started to identify where the code need to be changed and written TODOS

First rewrite how to run ramp, then incorporate @mohammadamint plotting scheme into the usecase, user and appliance classes, then modify cli for .xlsx users, then go through the docs and the jupyter notebook examples to adhere to the new unified way of running ramp

Bachibouzouk commented:
@mohammadamint  @FLomb, I edited the way the input files could be run, do you mind providing a WIP review before I go further into this direction?

Bachibouzouk commented on ramp/example/input_file_1.py:
```diff
@@ -254,3 +254,31 @@
+    from ramp.core.core import UseCase
+
+    uc = UseCase(
+        users=User_list,
```
@mohammadamint, @FLomb - here one can add `date_start` and or `date_end` to run simulations. test need to be written for the various possible case (no `date_start`but `date_end` and `num_days` provided, etc..)

mohammadamint replied:
@Bachibouzouk i am off grid with very poor internet connection! Sorry 😐

Bachibouzouk commented on ramp/core/core.py:
```diff
@@ -522,6 +739,16 @@ def generate_single_load_profile(
+            if self.usecase is None:
+                # logging warning
+                print(
+                    "You are generating ramp demand from a User not bounded to a UseCase instance, a default one has been created for you "
```
We can provide a better warning here letting the ramp user know about the UseCase class

Bachibouzouk commented on tests/test_notebooks.py:
```diff
@@ -0,0 +1,28 @@
+import os
```
This test help make sure the notebooks in the documentation are not broken. This is a smoke test, a single error raised upon execution of one cell in a notebook is enough to fail the test. Failed notebook are saved in the root of the repo for further inspection (to be deleted then)

FLomb reviewed (commented):
I like how the UseCase allows easily using the old inputs, but why are we putting all the operations on these inputs, such as generating and even plotting profiles, in the input files themselves?

In my opinion, input files should be exclusively about the inputs, shouldn't they? And all the operations on the inputs should happen elsewhere, e.g., in the `ramp_run.py` script.

FLomb commented on ramp/ramp_run.py:
```diff
@@ -58,5 +59,12 @@ def run_usecase(
     else:
-        raise TypeError("Only the .xlsx file format is supported for ramp command line")
+        raise TypeError(
+            "Only the .py and .xlsx file format are supported for ramp command line"
```
I am a bit confused by why here we say .py and .xlsx file formats are supported for ramp command line, while in the `cli.py` file we say that only .xlsx is supported for ramp command line.

FLomb commented:
I tested running the .py example files from the command line, and it works smoothly. Nice job! One minor comment: when the CLI prints the start and end times based on the given number of days, it includes decimal figures for the seconds, which are redundant. For example:
```
please indicate the number of days to be generated: 7
Please wait...
You will simulate 7 days from 2023-11-29 16:23:05.286940 until 2023-12-05 16:23:05.286940
```
Can we get rid of the decimal figures for the seconds?

FLomb commented:
> I like how the UseCase allows for the easy use of the old inputs. At first, I didn't get why all the operations on these inputs (plotting, etc.) were included in the input files, but now I get that this was the plan agreed in response to issue #83 as well as something allowing people to run via the command line. So that's good.
> 
> On the other hand, the fix to issue #83 does not seem to automatically solve also the very much related issue #81. I am now looking into that so that we may solve both at once with this PR.

So, it looks like the issue can still be easily fixed the way I suggested in issue #81. It's enough to just remove lines 84 and 85 that erroneously add the `ramp` prefix to the file path. In fact, adding the `ramp` prefix turns the input file path it into a ramp sub-module, which only works for the three pre-defined example input files, but which is definitely not what this function should be doing, which is pointing at physical, local files and converting them. Can we just remove those two lines please?

Bachibouzouk commented:
> So, it looks like the issue can still be easily fixed the way I suggested in issue #81. It's enough to just remove lines 84 and 85 that erroneously add the `ramp` prefix to the file path. In fact, adding the `ramp` prefix turns the input file path it into a ramp sub-module, which only works for the three pre-defined example input files, but which is definitely not what this function should be doing, which is pointing at physical, local files and converting them. Can we just remove those two lines please?

I also tested and I agree, I will remove two two lines :) 

Right now I am fixing the jupyter notebooks, this is interesting as it raises some questions about RAMP usage like "Should a RAMP user be able to just define an appliance and still generate a profile for this appliance without providing neither a User nor a Usecase", same question for User without Usecase.

I think it should be possible and we should make it as intuitive as possible for the RAMP user to use RAMP, providing some default values when the RAMP user did not provide information. The best would be that I prepare a list of those questions for next RAMP dev meeting so we can all voice our opinion on those questions, what do you think @FLomb @mohammadamint ?

FLomb commented:
> > So, it looks like the issue can still be easily fixed the way I suggested in issue #81. It's enough to just remove lines 84 and 85 that erroneously add the `ramp` prefix to the file path. In fact, adding the `ramp` prefix turns the input file path it into a ramp sub-module, which only works for the three pre-defined example input files, but which is definitely not what this function should be doing, which is pointing at physical, local files and converting them. Can we just remove those two lines please?
> 
> I also tested and I agree, I will remove two two lines :)
> 
> Right now I am fixing the jupyter notebooks, this is interesting as it raises some questions about RAMP usage like "Should a RAMP user be able to just define an appliance and still generate a profile for this appliance without providing neither a User nor a Usecase", same question for User without Usecase.
> 
> I think it should be possible and we should make it as intuitive as possible for the RAMP user to use RAMP, providing some default values when the RAMP user did not provide information. The best would be that I prepare a list of those questions for next RAMP dev meeting so we can all voice our opinion on those questions, what do you think @FLomb @mohammadamint ?

Sounds good to discuss this in the next meeting. On the other hand, it would have been nice to wrap up this PR in such a way as to release v0.5.0 and be able to send the paper/code for review to JOSS before the next meeting. But, of course, if the notebooks need fixing anyway for working with the new UseCase, this is something to address before making a release.

FLomb reviewed (commented):
I have updated, fixed (when needed) and refined all the Jupyter notebooks. The problem is that all the files in `docs/source/examples/`, which are essentially an `.rst` copy-paste of the notebooks, would also need equivalent updating (actually, even more updating because they are not even up to date with the state the notebooks wee before my changes).

I have already raised some time ago the issue that requiring this kind of double work any time we want to update the notebooks is very inefficient and tedious. Can we somehow automate the Jupyter-notebooks translation into .rst pages?

mohammadamint commented:
> I have updated, fixed (when needed) and refined all the Jupyter notebooks. The problem is that all the files in `docs/source/examples/`, which are essentially an `.rst` copy-paste of the notebooks, would also need equivalent updating (actually, even more updating because they are not even up to date with the state the notebooks wee before my changes).
> 
> I have already raised some time ago the issue that requiring this kind of double work any time we want to update the notebooks is very inefficient and tedious. Can we somehow automate the Jupyter-notebooks translation into .rst pages?

Rather than implementing all the changes both in rst files and jupyter files, we can download the modified jupyter files as rst to replace the docs examples. But there is a way to automatize this. I will open an issue to work on it.

Bachibouzouk commented on README.rst:
```diff
@@ -193,10 +203,10 @@ Building a model with a python script
 
    # importing functions
-   from ramp import User,calc_peak_time_range,yearly_pattern
+   from ramp import UseCase, User
```
Agree, I just translated the existing example but we should also showcase the `UseCase` more

mohammadamint commented:
> > I have updated, fixed (when needed) and refined all the Jupyter notebooks. The problem is that all the files in `docs/source/examples/`, which are essentially an `.rst` copy-paste of the notebooks, would also need equivalent updating (actually, even more updating because they are not even up to date with the state the notebooks wee before my changes).
> > I have already raised some time ago the issue that requiring this kind of double work any time we want to update the notebooks is very inefficient and tedious. Can we somehow automate the Jupyter-notebooks translation into .rst pages?
> 
> Rather than implementing all the changes both in rst files and jupyter files, we can download the modified jupyter files as rst to replace the docs examples. But there is a way to automatize this. I will open an issue to work on it.

@FLomb  I could manage this issue with a script. I will create a PR on top of these changes

Bachibouzouk commented on ramp/core/core.py:
And I also see that older functionalities are present in the docs, there should certainly be a command to automatize the building of API reference from source code. How did you build it inititally @mohammadamint ?

mohammadamint replied:
I build them by manually adding including the functions in the toctree. As there are some functions that are private, i think we should avoid including all of them, just put the one that the user might interact with

Bachibouzouk commented:
> @Bachibouzouk What do you think? I am using autoclass to automatically include all the functions documented.

Yes I agree with this, there is not many methods to document anyways, it is quite quick to do it manually :)

Bachibouzouk commented:
> I can take care of this, but I see some new functions also that dont have the docstrings yet.

@mohammadamint - If you take care of this (in a separate PR pointing to `fix/unify-usecase`, or after @FLomb edited the README), then I will edit the docstrings tonight and we could merge it tomorrow

mohammadamint commented on docs/source/api_references.rst:
```diff
@@ -65,7 +75,6 @@ Utilities
     :toctree: api_document/
 
     yearly_pattern
```
Done!

Bachibouzouk commented:
I recreated the plot_class rst file from the .ipynb by explicitely removing the cells with plotly figure because the size of the file was 10MB. I replaced them by the png I created manually. As the Plot_class.rst was already in git history I am not sure I changed much the size of the repo and we can also reverse to use directly the .rst with plotly figures embedded (this is heavy because all the data for the figures are present in the html as well as the plotly js code so that the figures are interactive.

Bachibouzouk commented:
@FLomb @mohammadamint maybe you can have a look at my latest changes to the documentation and my addendum to the changelog, and also to this comment: https://github.com/RAMP-project/RAMP/pull/102#issuecomment-1839604314

Then we can finally merge, what is not perfect can be corrected into subsequent PRs, I think this one is already big enough now ^^

FLomb reviewed (approved):
All seems to be in good order now, so I'll approve the PR
